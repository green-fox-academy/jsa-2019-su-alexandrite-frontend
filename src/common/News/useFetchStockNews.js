import { useEffect, useState } from 'react';
import {
  NEWS_API_URL,
  NEWS_API_KEY,
} from 'react-native-dotenv';

const fetchNews = async (query, page, pageSize) => {
  const newsUrl = new URL(`${NEWS_API_URL}/everything`);
  const parsedQuery = (typeof query.join) === 'function'
    ? query.join(' OR ')
    : query;
  newsUrl.searchParams.append('apiKey', NEWS_API_KEY);
  newsUrl.searchParams.append('q', parsedQuery);
  newsUrl.searchParams.append('pageSize', pageSize);
  newsUrl.searchParams.append('page', page);
  newsUrl.searchParams.append('sortBy', 'publishedAt');
  newsUrl.searchParams.append('language', 'en');
  const res = await fetch(newsUrl);
  if (!res.ok) {
    switch (res.status) {
      case 404:
        throw new Error(`Cannot find any news about ${query}`);
      case 426:
        throw new Error(`No more news about ${query}, please come back later!`);
      default:
        throw new Error('Oops, there\'s something wrong with our app.');
    }
  }
  const json = await res.json();
  return json.articles.map(({
    publishedAt, title, urlToImage, url,
  }) => ({
    datetime: publishedAt,
    headline: title,
    image: urlToImage,
    url,
    id: (new Date() * Math.random()).toString(),
  }));
};

export default (query, page = 1, pageSize = 20) => {
  const [news, setNews] = useState([]);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    const fetchData = async () => {
      try {
        const result = await fetchNews(query, page, pageSize);
        setNews(page === 1 ? result : [...news, ...result]);
        setIsLoading(false);
      } catch (err) {
        setError(err);
        setIsLoading(false);
      }
    };
    fetchData();
  }, [page]);
  return {
    news, error, isLoading,
  };
};
