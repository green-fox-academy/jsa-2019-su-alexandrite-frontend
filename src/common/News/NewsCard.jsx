import React from 'react';
import {
  View,
  ActivityIndicator,
  Text,
  TouchableHighlight,
  ViewPropTypes,
} from 'react-native';
import { useNavigation } from 'react-navigation-hooks';
import PropTypes from 'prop-types';
import useFetchStockNews from './useFetchStockNews';
import styles from '../styles';
import Card from '../Card';
import ErrorMessage from '../ErrorMessage';
import NewsMappedList from './NewsMappedList';
import CardFooter from '../CardFooter';
import Row from '../Row';

const NewsCard = ({ query, style }) => {
  const { news, isLoading, error } = useFetchStockNews(query, 1, 2);
  const navigator = useNavigation();
  return (
    <Card title="News" style={style}>
      {!error ? (
        <>
          {isLoading && <ActivityIndicator size="large" />}
          {!isLoading && news && (
            <NewsMappedList news={news} />
          )}
          <CardFooter>
            <TouchableHighlight
              style={styles.cardFooterButton}
              onPress={() => navigator.navigate('NewsList', { query })}
              underlayColor="#eee"
            >
              <Text style={styles.cardFooterButtonText}>View more</Text>
            </TouchableHighlight>
          </CardFooter>
        </>
      ) : <ErrorMessage message={error.message} />}
    </Card>
  );
};

NewsCard.propTypes = {
  query: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]).isRequired,
  style: ViewPropTypes.style,
};

NewsCard.defaultProps = {
  style: null,
};

export default NewsCard;
