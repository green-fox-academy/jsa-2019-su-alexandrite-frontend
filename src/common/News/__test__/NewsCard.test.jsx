import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import useFetchStockNews from '../useFetchStockNews';
import NewsCard from '../NewsCard';

jest.mock('react-redux');
jest.mock('react-navigation-hooks');
jest.mock('../NewsMappedList', () => '<NewsMappedList />');
jest.mock('../useFetchStockNews');

describe('<Popup />', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useFetchStockNews.mockReturnValue({
      news: [],
      isLoading: false,
      error: null,
    });
    useNavigation.mockReturnValue({
      push: jest.fn(),
      navigate: jest.fn(),
    });
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Should render properly and match the snapshot', () => {
    const tree = renderer.create(
      <NewsCard query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render properly and match the snapshot on error', () => {
    useFetchStockNews.mockReturnValueOnce({
      news: null,
      isLoading: false,
      error: {
        message: 'test error',
      },
    });
    const tree = renderer.create(
      <NewsCard query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render properly and match the snapshot when loading', () => {
    useFetchStockNews.mockReturnValueOnce({
      news: null,
      isLoading: true,
      error: null,
    });
    const tree = renderer.create(
      <NewsCard query="test" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should navigate to NewsList with query passed', () => {
    const mockNavigate = jest.fn();
    useNavigation.mockReturnValueOnce({
      navigate: mockNavigate,
    });
    const wrapper = shallow(
      <NewsCard query="test" />,
    );
    wrapper.find('TouchableHighlight').first().simulate('press');
    expect(mockNavigate).toHaveBeenCalledWith('NewsList', { query: 'test' });
  });
});
