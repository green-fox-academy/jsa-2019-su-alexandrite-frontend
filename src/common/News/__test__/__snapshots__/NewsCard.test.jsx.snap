// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> Should render properly and match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 15,
        "padding": 15,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      null,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "#333",
        "fontSize": 13,
        "fontWeight": "600",
        "lineHeight": 18,
        "marginBottom": 15,
      }
    }
  >
    News
  </Text>
  <View
    style={
      Object {
        "width": 690,
      }
    }
  >
    <<NewsMappedList />
      news={Array []}
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "paddingTop": 15,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "borderRadius": 3,
            "padding": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 12,
            }
          }
        >
          View more
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Popup /> Should render properly and match the snapshot on error 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 15,
        "padding": 15,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      null,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "#333",
        "fontSize": 13,
        "fontWeight": "600",
        "lineHeight": 18,
        "marginBottom": 15,
      }
    }
  >
    News
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#fff6f6",
        "borderRadius": 6,
        "paddingHorizontal": 8,
        "paddingVertical": 15,
        "width": 690,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#9f3a38",
          "fontWeight": "bold",
        }
      }
    >
      test error
    </Text>
  </View>
</View>
`;

exports[`<Popup /> Should render properly and match the snapshot when loading 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 6,
        "elevation": 3,
        "marginBottom": 15,
        "padding": 15,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "x": 0,
          "y": 4,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 6,
      },
      null,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "#333",
        "fontSize": 13,
        "fontWeight": "600",
        "lineHeight": 18,
        "marginBottom": 15,
      }
    }
  >
    News
  </Text>
  <View
    style={
      Object {
        "width": 690,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
